### PR TITLE
User policy support for temp keys

### PR DESCRIPTION
**Pull request by chenriksson**

### CODE REVIEW

**Comment by skofman1** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

I guess this is not supposed to be here...

---
Refers to: src/NuGetGallery/App_Data/Files/package-backups/chenriks/2.0.0/1cBXdBjDcqTgEnhvoAz5YdvfXiIOGlErkduAID7kkfl3rCdrvkniZsQwmoO745LTo1534-XaP6iLJXsNVfpj8A2..nupkg:1 in 796a739. [](commit_id = 796a739322e680d4478aff0c63dd92efd5b4d199, deletion_comment = False)

File: src/NuGetGallery/ExtensionMethods.cs [hunk 1]
```diff
@@ -407,10 +407,30 @@ public static bool HasVerifyScope(this IIdentity self)
...
+
+        /// <summary>
+        /// Get the credential used for authentication by the current user.
+        /// </summary>
+        /// <param name="user">Current user.</param>
+        /// <param name="identity">Authenticated claims identity.</param>
+        /// <returns>Current credential.</returns>
+        public static Credential GetCurrentCredential(this User user, IIdentity identity)
+        {
+            var apiKey = ((ClaimsIdentity)identity).GetClaimOrDefault(NuGetClaims.ApiKey);
```
Comment (skofman1) [inline src/NuGetGallery/ExtensionMethods.cs]: identity might not be ClaimsIdentity, and this will throw

File: src/NuGetGallery/ExtensionMethods.cs [hunk 2]
```diff
@@ -407,10 +407,30 @@ public static bool HasVerifyScope(this IIdentity self)
         {
             var scopeClaim = self.GetScopeClaim();
 
-            return ScopeEvaluator.ScopeClaimsAllowsActionForSubject(scopeClaim, subject: null,
+            return !ScopeEvaluator.IsEmptyScopeClaim(scopeClaim) &&
+                ScopeEvaluator.ScopeClaimsAllowsActionForSubject(scopeClaim, subject: null,
```
Comment (skofman1) [inline src/NuGetGallery/ExtensionMethods.cs]: I know this is a small fix, but it's hard to code review as part of such a big change. Can this be extracted to a separate PR?

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 3]
```diff
@@ -239,23 +236,30 @@ private Credential GetCurrentCredential(User user)
...
+
             TelemetryService.TrackVerifyPackageKeyEvent(id, version, user, User.Identity, result?.StatusCode ?? 200);
 
             return (ActionResult)result ?? new EmptyResult();
         }
 
-        private HttpStatusCodeWithBodyResult VerifyPackageKeyInternal(User user, Credential credential, string id, string version)
+        private async Task<HttpStatusCodeWithBodyResult> VerifyPackageKeyInternal(User user, Credential credential, string id, string version)
         {
+            // Evaluate user security policy requirements.
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: why do we need security policy for verification?

File: src/NuGetGallery/Extensions/HttpRequestExtensions.cs [hunk 4]
```diff
@@ -11,6 +12,28 @@ namespace NuGetGallery
...
+        {
+            return GetNuGetClientVersion(new HttpRequestWrapper(request));
+        }
+
+        private static string GetNuGetClientVersion(this HttpRequestBase request)
+        {
+            return request?.Headers[Constants.ClientVersionHeaderName];
+        }
+
+        public static bool IsNuGetClientVersionOrHigher(this HttpRequestBase request, Version requiredVersion)
```
Comment (skofman1) [inline src/NuGetGallery/Extensions/HttpRequestExtensions.cs]: is there a better spot for this method.. not really related to HttpRequestExtensions..

File: src/NuGetGallery/Services/SecurityPolicyResult.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery
+{
+    public class SecurityPolicyResult
```
Comment (skofman1) [inline src/NuGetGallery/Services/SecurityPolicyResult.cs]: please add documentation

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 6]
```diff
@@ -364,6 +368,16 @@ public virtual Task<ActionResult> CreatePackagePost()
                                 errorsString));
                         }
 
+                        var packageId = nuspec.GetId();
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: the scenario when a user tries to push a package with the wrong client version  should break much faster then here. I recommend adding the verification in the attribute..

File: src/NuGetGallery/ExtensionMethods.cs (hunk 2 above)
Comment (chenriksson) [inline src/NuGetGallery/ExtensionMethods.cs]: Done - will update this PR to exclude.

File: tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs [hunk 7]
```diff
@@ -88,6 +88,9 @@ public TestableApiController(MockBehavior behavior = MockBehavior.Default)
                     return Task.FromResult(package);
                 });
 
+
+            SecurityPolicyService = new SecurityPolicyService(EntitiesContext, AuditingService, new Diagnostics.DiagnosticsService());
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs]: The pattern of these tests is that everything should be mocked. Creating a normal `SecurityPolicyService` here does not fit that. The `TestableApiController` should declare a `Mock<SecurityPolicyService>` and should `Setup` and `Verify` it when specific functionality is being tested.

The tests here should
1 - Verify that the `SecurityPolicyService` is called with the expected input.
2 - Return the expected output of the expected input.

Tests of the actual functionality of `SecurityPolicyService` in `SecurityPolicyServiceFacts` and should 
1 - Call `SecurityPolicyService` with the expected input
2 - Expect that `SecurityPolicyService` returns the expected output.

File: tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs [hunk 8]
```diff
@@ -359,6 +362,109 @@ public class TheCreatePackageAction
...
+                packageRegistration.Id = "theId";
+                packageRegistration.Owners.Add(user);
+                var package = new Package();
+                package.PackageRegistration = packageRegistration;
+                package.Version = "1.0.42";
+                packageRegistration.Packages.Add(package);
+
+                var controller = new TestableApiController();
+                controller.SetCurrentUser(user);
+                controller.MockPackageService.Setup(p => p.FindPackageRegistrationById(It.IsAny<string>()))
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs]: I don't see `controller.Request.Headers[Constants.ClientVersionHeaderName]` set here even though the test is named `WritesAuditRecordAndReturns400IfSecurityPolicyAndLowerClientVersion`.

File: tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs [hunk 9]
```diff
@@ -359,6 +362,109 @@ public class TheCreatePackageAction
...
+                ResultAssert.IsStatusCode(result, HttpStatusCode.BadRequest);
+
+                Assert.True(controller.AuditingService.WroteRecord<PackageVerificationKeysPolicyAuditRecord>(ar =>
+                    ar.Action == UserSecurityPolicyAuditAction.PackageVerificationKeysPolicy_CreatePackage
+                    && ar.PushedPackage.Id == package.PackageRegistration.Id
+                    && ar.PushedPackage.Version == package.Version));
+            }
+
+            [Fact]
+            public async Task WritesAuditRecordAndReturns201IfSecurityPolicyAndClientVersionMet()
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs]: These three tests are very similar, the only difference being how `controller.Request.Headers[Constants.ClientVersionHeaderName]` is set. Can we extract the shared code and make these tests simpler?

File: tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs [hunk 10]
```diff
@@ -1270,6 +1405,8 @@ public class TheVerifyPackageKeyAction : PackageVerificationKeyContainer
                     Version = "1.0.0"
                 };
                 var controller = SetupController(CredentialTypes.ApiKey.VerifyV1, scope, package);
+                var user = controller.OwinContext.GetCurrentUser();
+                user.SecurityPolicies.Add(new PackageVerificationKeysPolicy());
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/ApiControllerFacts.cs]: I would split this into two tests
1 - `VerifyPackageKeyReturns200_ApiKeyVerifyV1` (the test before you modified it)
2 - `VerifyPackageKeyReturns200_ApiKeyVerifyV1_WithPackageVerificationKeysPolicy`

We need to make sure both scenarios are unaffected.

File: tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+                // Assert
+                Assert.False(service.Auditing.WroteRecord<PackageVerificationKeysPolicyAuditRecord>(ar =>
+                    ar.Action == UserSecurityPolicyAuditAction.PackageVerificationKeysPolicy_CreatePackage));
+            }
+
+            [Fact]
+            public async Task ReturnsSuccessIfUserNotSubscribed()
+            {
+                var httpContext = CreateHttpContext(clientVersion: "1.0.0");
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs]: Should test what happens when the user is not subscribed and the client version is met as well.

File: tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+            private SecurityPolicyService CreateService()
+            {
+                return new SecurityPolicyService(Get<IEntitiesContext>(), Get<IAuditingService>(), new DiagnosticsService());
+            }
+        }
+
+        public class TheCanCreatePackageMethod : TestContainer
+        {
+            [Fact]
+            public async Task WritesAuditRecordIfUserSubscribed()
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs]: `IfUserSubscribedAndClientVersionNotMet`

File: tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+                Assert.True(service.Auditing.WroteRecord<PackageVerificationKeysPolicyAuditRecord>(ar =>
+                    ar.Action == UserSecurityPolicyAuditAction.PackageVerificationKeysPolicy_CreatePackage &&
+                    ar.Username == "a" &&
+                    ar.PushedPackage.Id == "b" && ar.PushedPackage.Version == "1.2.3" &&
+                    // Writes error due to client version 1.0.0.
+                    !string.IsNullOrEmpty(ar.ErrorMessage)));
+            }
+
+            [Fact]
+            public async Task DoesNotWriteAuditRecordIfUserNotSubscribed()
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs]: Should test both client version met and client version not met for this.

File: tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            {
+                // Arrange
+                var service = CreateService();
+                var user = new User("a");
+                user.Credentials.Add(new Credential(CredentialTypes.ApiKey.V2, Guid.Empty.ToString()));
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs]: Need to check what happens when the policy is satisfied as well.

File: tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+                    ar.Action == UserSecurityPolicyAuditAction.PackageVerificationKeysPolicy_VerifyPackageKey));
+            }
+
+            [Fact]
+            public async Task ReturnsSuccessIfUserNotSubscribed()
+            {
+                // Arrange
+                var service = CreateService();
+                var user = new User("a");
+                user.Credentials.Add(new Credential(CredentialTypes.ApiKey.V2, Guid.Empty.ToString()));
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Services/SecurityPolicyServiceFacts.cs]: Need to check what happens when the policy is satisfied as well.

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs [hunk 16]
```diff
@@ -101,6 +101,11 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
 
             modelBuilder.Entity<PackageLicense>()
                 .HasKey(l => l.Key);
+            
+            modelBuilder.Entity<SecurityPolicy>()
```
Comment (skofman1) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: IMO the different security policy types shouldn't be part of the data model. They don't affect the DB schema.
We have similar situation with credentials for example: there are different types of credentials, but those are not mentioned in EntitiesContext, and other layers handle the different credential types.

File: src/NuGetGallery.Core/Entities/PackageVerificationKeysPolicy.cs [hunk 17]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGetGallery
+{
+    public class PackageVerificationKeysPolicy
```
Comment (skofman1) [inline src/NuGetGallery.Core/Entities/PackageVerificationKeysPolicy.cs]: we need to discuss and decide how security policies are defined and named.
For example: PackageVerificationKeysPolicy is used for both Verify API and Push API. This is confusing. 
A more general approach would be, to define two policies:
1. MinClientVersionForPush  (with version parameter)
2. EnforceVerificationAPIKeys  (true/false)

Also, the client version we require shouldn't be part of the policy (cause this way we can't reuse it), but should be part of the onboarding logic.

File: src/NuGetGallery/Services/SecurityPolicyService.cs [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+            // On subscribing, this policy requires expiration of API keys with push capabilities.
+            var expiredCount = 0;
+            foreach (var credential in user.Credentials)
+            {
+                if (credential.IsApiKeyWithPushCapability())
+                {
+                    await Auditing.SaveAuditRecordAsync(
+                        new UserAuditRecord(user, AuditedUserAction.ExpireCredential, credential));
+
+                    // Provide 1-week window before expiration to allow setup of new keys.
```
Comment (skofman1) [inline src/NuGetGallery/Services/SecurityPolicyService.cs]: The onboarding logic is used only by admins, and belongs in the admin area: NuGetGallery/areas/admin.

Comment (chenriksson) [inline src/NuGetGallery/Services/SecurityPolicyService.cs]: Will do onboarding in separate PR

**Comment by chenriksson** [comment]
@skofman1 @scottbommarito Thanks for the feedback... fed it into new design. Will open different PR for changes, and separate PRs for onboarding and auditing.
